### PR TITLE
[StaticQuant] Update how block_size is calculated with Observers

### PR DESCRIPTION
**Pull request by drisspg**
Stacked PRs:
 * #807
 * __->__#815

--- --- ---

[StaticQuant] Update how block_size is calculated with Observers

### CODE REVIEW

File: torchao/quantization/observer.py [hunk 1]
```diff
@@ -16,14 +16,35 @@
...
+    """
+    Represents per-axis granularity in quantization.
+
+    This granularity type calcualtes different quantization parameters
+    along a specified axis of the tensor.
+
+    Attributes:
+        axis (int): The axis along which reduction is performed.
+    """
     axis: int
```
Comment (drisspg) [inline torchao/quantization/observer.py:48]: @jerryzh168 
This kinda works backwards to what I would think

e.g with tensor:
9x2048

If I was calculating with axis wise and axis = 0, my first thought would be that this would produce min_values
1x2048 e.g. along the columns and vice versa  for axis=1 but it seems that this is flipped

Comment (jerryzh168) [reply torchao/quantization/observer.py:48]: axis = 0 means one pair of qparams per element in axis = 0, in this case it would be 9 qparams.

this is consistent with the `ch_axis` arg we have before in: https://github.com/pytorch/pytorch/blob/f2a7228aed15f97cd4389b4240fa24e25b5eeb4f/torch/ao/quantization/observer.py#L690 and https://github.com/pytorch/pytorch/blob/7b280c31bac252c7741067a68b1675aefeb1da35/aten/src/ATen/native/native_functions.yaml#L7539

File: torchao/quantization/observer.py [hunk 2]
```diff
@@ -130,8 +160,13 @@ def forward(self, input: torch.Tensor):
...
 
         input_detached = input.detach()
-        if self.block_size is None:
-            self.block_size = get_block_size(input_detached.shape, self.granularity_type)
+        assert self.granularity_type is not None, "granularity_type is None"
+        block_size = get_block_size(input_detached.shape, self.granularity_type)
+
+        # If we are doing PerTensor quantization then we do not need to reduce along a given
+        # Dimension. The input tensor can vary sizes and we dont want to cache a block_size
+        if self.block_size is None and not isinstance(self.granularity_type, PerTensor):
```
Comment (jerryzh168) [inline torchao/quantization/observer.py:168]: I think it probably makes sense to remove `self.block_size` and always recalculate the block_size

**Review by jerryzh168 — APPROVED** [review]
looks great, thanks for adding docs and the additional test! had one comment inline, please address the comment before merging

File: torchao/quantization/observer.py (hunk 2 above)
Comment (drisspg) [reply torchao/quantization/observer.py:168]: you are right I thought it was needed in choose_qparams, but we are good
